runtime/doc/quickfix.txt

```diff
@@ -1335,6 +1335,59 @@ g:javac_makeprg_params variable.  For example: >
 
 	let g:javac_makeprg_params = "-Xlint:all -encoding utf-8"
 <
+SPOTBUGS						*compiler-spotbugs*
+
+SpotBugs is a static analysis tool that can be used to find bugs in Java.
+It scans the Java bytecode of all classes in the currently open buffer.
+
+Commonly used compiler options can be added to 'makeprg' by setting the
+b/g:spotbugs_makeprg_params variable.  For example: >
+
+	let b:spotbugs_makeprg_params = "-longBugCodes -effort:max -low"
+
+The global default is "-workHard -experimental".
+
+By default, the class files are searched in the directory where the source
+files are placed.  However, typical Java projects use distinct directories
+for source files and class files.  To make both known to SpotBugs, assign
+their paths (relative to the current directory) to the following two
+properties (using the example of a typical Maven project): >
+
+	let g:spotbugs_properties = {'sourceDirName': 'src/main/java',
+				   \ 'classDirName':  'target/classes'}
+
+Pre- and post-compiler actions, to first refresh the class files for the
+current buffer, and then run `spotbugs` on them, can be enabled by adding
+to your |vimrc| (before any |FileType| event): >
+
+	let g:spotbugs_properties = {
+		\ 'PreCompilerAction':  function('CompilerJavacMake'),
+		\ 'PostCompilerAction': function('SpotBugsMake'),
+	\}
+
+	function! CompilerJavacMake() abort
+		if exists('b:spotbugs_class_files')
+			for pathname in b:spotbugs_class_files
+				call delete(eval(pathname))
+			endfor
+
+			let b:spotbugs_class_files = []
+		endif
+
+  		compiler javac
+  		exe 'make -Xlint -Xdiags:verbose'
+			\ '-d' shellescape(get(g:, 'spotbugs_properties.classDirName', '.'), 1) '%:S'
```

Comment:
And so in this example, the `-d` pathname argument should use   
only whatever `g:javac_makeprg_params` can offer:

```vim
		compiler javac
		exe 'make -Xlint -Xdiags:verbose'
			\ '-d' shellescape(get(g:, 'spotbugs_properties.classDirName', '.'), 1) '%:S'
```

Also, the `get()` syntax looks suspicious.  Two dictionaries  
are involved: `g:` and `spotbugs_properties`.  Try instead:

```vim
echo shellescape(get(get(g:, 'spotbugs_properties', {}), 'sourceDirName', '.'), 1)
```

Reply:
> Also, the get() syntax looks suspicious.

Full agreement; fixed

Reply:
> And so in this example, the -d pathname argument should use
only whatever g:javac_makeprg_params can offer:

This conclusion is not clear to me. Assuming `classDirName` (by the way, rather `Path`?) is relative to their common root dir, shouldn't this be prepended?

Reply:
So rather a reverse of 

```vim
" runtime/compiler/spotbugs.vim (lines 87-90)
" Substitute "bin_dir_name" for the rightmost "src_dir_name"
let candidate_type_name = strpart(a:src_type_name, 0, tail_idx)..
    \ bin_dir_name..
    \ strpart(a:src_type_name, (tail_idx + strlen(src_dir_name)))
```

for `spotbugs_properties.classDirName` using `%:p`.

Reply:
The pathname argument passed for the `-d` (and `-cp`) options  
of Javac and the value of `classDirName` are independent of  
each other.

As is shown [above](https://github.com/vim/vim/pull/16001#discussion_r1844591894), `classDirName` and `sourceDirName` must  
contain distinct pathname components (but can redundantly  
contain common pathname components too) _between_ the absolute  
pathname of a Java source file and the absolute name of its  
class file, if compiled.  Whereas the argument for `-d` is  
usually made relative to the current working directory (but  
can also be absolute) and it points to the directory for  
class files to be stored in.

I notice that when SpotBugs attempts to jump to an offending  
line from its report, it always insists on loading the file  
matching the pathname parsed from (`%f`).  This means that    
you have to `:cd` to the root package directory and build  
your relative path for `-d` from there.  

For example:

```shell
cd /tmp/project
VIMRUNTIME=/path/to/fork/runtime vim pom.xml \
    $(find src/main/java -iname \*.java)
```

```vim
:cd src/main/java
:let g:spotbugs_properties = {
	\ 'sourceDirName': '/tmp/project/src/main/java/com/acme/',
	\ 'classDirName': '/tmp/project/target/classes/com/acme/'}
:let g:javac_makeprg_params = '-d ../../../target/classes -cp ../../../target/classes'

:compiler javac
:make %:S
:compiler spotbugs
:make %:S
```

The values are clearly different.

Reply:
> I notice that when SpotBugs attempts to jump to an offending
line from its report, it always insists on loading the file
matching the pathname parsed from (%f). This means that
you have to :cd to the root package directory and build
your relative path for -d from there.

To get the full path in its output for Vim's quickfix parser, while avoiding these directory changes, is the point of Spotbug's `-sourcepath`

Reply:
Then it fails at it for packages of varying directory depths  
unless your current working directory is relative to `%f`.  
The current resolved value for `-sourcepath` only removes the  
source file, is this correct for packaged types?

For example (`cd /tmp/project`):

```
(%:p:h): src/com/acme/Test.java -> [New DIRECTORY]
(%:p:h): src/com/Test.java -> [New DIRECTORY]
(%:p:h): src/Test.java
```

whereas `%f` would contain:

```
com/acme/Test.java:x:y Trouble here
com/Test.java:x:y Trouble here
/tmp/project/src/Test.java:x:y Trouble here
```

After some experimenting, these values appear to work for  
`-sourcepath` when `:pwd` is at `/tmp/project` or `/tmp`:

```
(%:p:h:h:h): src/com/acme/Test.java
(%:p:h:h): src/com/Test.java
(%:p:h): src/Test.java
```

whereas `%f` would now contain:

```
/tmp/project/src/com/acme/Test.java:x:y Trouble here
/tmp/project/src/com/Test.java:x:y Trouble here
/tmp/project/src/Test.java:x:y Trouble here
```

My current workaround for a mismatch between the expected  
current working directory, derived from parsing `%f`, and  
the actual working directory:

```vim
function! s:SpotBugsMake() abort
	make %:S

	" Make the current working directory match the expected one from "%f".
	if wordcount().bytes == 0
		let new_buffer = bufname()
		execute 'buffer ' . expand('#')
		execute 'bwipeout ' . new_buffer
		execute 'cd ' . expand('%:p:s?' . new_buffer . '??')
		" FIXME: Can ":cc" be recovered without another ":make"?
		make %:S
	endif
endfunction

let g:spotbugs_properties = {
	\ 'PostCompilerAction': function('s:SpotBugsMake'),
\ }
```

Reply:
See https://github.com/vim/vim/pull/16001#issuecomment-2483534732